File: Go/problem012.go
package projecteuler

// #12 Highly divisible triangular number - Project Euler
// http://projecteuler.net/index.php?section=problems&id=12
import (
	"log"
	"projecteuler/my"
)

func P012() {
	x := 0
	for i := 1; ; i++ {
		x += i
		fs := my.Factors(x)
		if len(fs) > 500 {
			log.Println(x)
			return
		}
	}
}

File: Zig/problem012.zig
// zig version 0.11.0-dev
const std = @import("std");
const stdout = std.io.getStdOut().writer();

// #12 Highly divisible triangular number - Project Euler
// http://projecteuler.net/index.php?section=problems&id=12
fn countDivisors(x: u32) usize {
    var count: usize = 0;
    var i: u32 = 1;

    while (i * i <= x) : (i += 1) {
        if (x % i == 0) {
            count += 1;
            if (!(x / i == i or i == x)) {
                count += 1;
            }
        }
    }

    return count;
}

pub fn main() !void {
    var triangle: u32 = 1;
    var adder: u32 = 2;

    while (countDivisors(triangle) <= 500) : (adder += 1) {
        triangle += adder;
    }

    try stdout.print("{}\n", .{triangle});
}
